tezos/context/src/gc/worker.rs

```diff
@@ -400,42 +609,308 @@ impl GCThread {
-            self.objects_generation
-                .entry(hash_id)?
-                .replace(current_generation);
+            let generation = self.objects_generation.entry(hash_id)?;
+            *generation = self.current_generation;
         }
 
-        // Recursively update generation of all the children
-        for hash_id in hash_ids {
-            let hash_id = match hash_id {
-                Some(hash_id) => hash_id,
-                None => break,
-            };
+        for index in start..end {
+            let hash_id = hash_ids[index];
             self.traverse_and_mark_tree(
                 hash_id,
-                current_generation,
+                hash_ids,
                 nobjects,
-                depth + 1,
+                depth,
                 max_depth,
                 objets_total_bytes,
             )?;
         }
 
+        assert_eq!(hash_ids.len(), end);
+        hash_ids.remove_last_nelems(end - start);
+
+        Ok(())
+    }
+
+    fn copy_hash_to_snapshot(
+        &self,
+        hash_id: HashId,
+        on_disk_repo: &mut Persistent,
+        hash: &mut ObjectHash,
+    ) -> Result<HashId, GCError> {
+        {
+            // Copy to object hash (found in the in-mem context) into `hash`
+            // Lock the in-mem repo as short as possible
+            let in_mem_repo = self.in_mem_repository()?.read();
+            let object_hash = in_mem_repo.get_hash(ObjectReference::new(Some(hash_id), None))?;
+            *hash = *object_hash;
+        }
+
+        // Copy the hash into the on-disk repository, it now has a new `HashId`
+        let hash_id = on_disk_repo
+            .get_vacant_object_hash()?
+            .write_with(|entry| entry.copy_from_slice(hash))?;
+
+        Ok(hash_id)
+    }
+
+    fn in_mem_repository(&self) -> Result<&Arc<RwLock<ContextKeyValueStore>>, GCError> {
+        self.in_mem_repo
+            .as_ref()
+            .ok_or(GCError::InMemContextNotFound)
+    }
+
+    fn maybe_write_to_disk(
+        &self,
+        output: &mut SerializeOutput,
+        on_disk_repo: &mut Persistent,
+    ) -> Result<(), GCError> {
+        let data_len = output.len();
+        let hashes_len = on_disk_repo.hashes_in_memory_len();
+        let strings_len = on_disk_repo
+            .string_interner
+            .new_bytes_since_last_serialize();
+
+        let data_limit_reached = data_len >= 20_000_000;
+        let hashes_limit_reached = hashes_len >= 625_000;
+        let strings_limit_reached = strings_len >= 10_000_000;
+
+        if data_limit_reached || hashes_limit_reached || strings_limit_reached {
+            on_disk_repo.commit_to_disk(output).map_err(DBError::from)?;
+            on_disk_repo.set_is_commiting();
+            on_disk_repo.deallocate_strings_shapes();
+            output.clear();
+        }
+
         Ok(())
     }
 
+    fn find_frozen_commit(&self, commit_hash_id: HashId, bytes: &mut Vec<u8>) -> Option<HashId> {
```

Comment:
What is a frozen commit?

Reply:
probably he means commit for cemented block maybe. Blocks that are <= HEAD~2.

Reply:
Yes that's a commit at the current block level minus 3. I will add some documentation